src/main/java/com/google/sps/servlets/ChallengesServlet.java

```diff
@@ -92,16 +101,69 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
     response.getWriter().println(json);
   }
 
+  /* add new challenge id to user status map, update current challenge id
+     and add challenge id to user completed challenge map */
+  @Override
+  public void doPut(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String completed_challenge_id = request.getParameter(COMPLETED_CHALLENGES);
+    String current_challenge_id = request.getParameter(CURRENT_CHALLENGE);
+    Payload payload = GoogleIdHelper.verifyId(request);
+    if (payload == null) {
+        response.setStatus(400);
+        return;
+    }
+
+    String user_id = payload.getSubject();
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    Query query = new Query(User.DATA_TYPE).setFilter(new FilterPredicate(User.ID, FilterOperator.EQUAL, user_id));
+    Entity entity = datastore.prepare(query).asSingleEntity();
+    User user = User.convertEntitytoUser(entity, user_id);
+
+    if(completed_challenge_id != null){
+      try{
+        updateUserChallenges(user, completed_challenge_id, current_challenge_id);
+        } catch(Exception e) {
+          System.err.println(e.getMessage());
+      }
+    }
+    datastore.put(user.toEntity());
+
+    response.setContentType("application/json; charset=UTF-8");
+    response.getWriter().println(user.toJSON());
+  }
+
   /** Converts array of challenges into json format */
-  private String convertToJsonUsingGson(ArrayList challenge_list){
+  private String convertToJsonUsingGson(ArrayList challenge_list) {
     String json = new Gson().toJson(challenge_list);
     return json;
   }
 
+  //This function Update's User's Current, and Completed challenges
+  // along with statuses
+  private void updateUserChallenges(User user, String compl_id, String cur_id) {
+    HashMap<String, Integer> challenge_statuses = user.getChallengeStatuses();
+    HashSet<String> completed_challenges = user.getCompletedChallenges();
+    user.setCurrentChallenge(cur_id);
+
+    //challenge_statuses.remove(compl_id);
+    user.appendToCompletedChallenges(compl_id);
+    
+    //add new challenge to challenge status.
+    for(String challenge_id : ChallengeData.CHALLENGES_MAP.keySet()){
```

Comment:
It seems like you're initializing user's challenge_status for all challenges in CHALLENGES_MAP. Wouldn't it make more sense to put that code into where the User is initialized? I.e. https://github.com/googleinterns/step72-2020/blob/e0fbc902167e97e4ba1699cb05c7b04d6414bdea/src/main/java/com/google/sps/servlets/UserServlet.java#L132 should read `.setChallengeStatuses(ChallengeData.DEFAULT_CHALLENGES_AND_STATUS)` where `ChallangeData.DEFAULT_CHALLENGES_AND_STATUS` should be similar to https://github.com/googleinterns/step72-2020/blob/e0fbc902167e97e4ba1699cb05c7b04d6414bdea/src/main/java/com/google/sps/servlets/UserServlet.java#L255 but modified to include all challenges present in `ChallangeData.CHALLENGES_MAP`.